ba/js/lesson4/script.js

```diff
@@ -0,0 +1,123 @@
+//Task 1
+
+const greaterThan = (numA, numB) => (numA > numB ? 1 : numA == numB ? 0 : -1);
+
+console.log(greaterThan(96, 6));
+console.log(greaterThan(0, 6));
+console.log(greaterThan(-4, -4));
+
+//Task 2
+
+const factorial = (n) => (n == 0 || n == 1 ? 1 : n * factorial(n - 1));
+
+console.log(factorial(6));
+
+//Task 3
+
+const threeNumbers = (num1, num2, num3) => +(num1.toString() + num2 + num3);
+
+console.log(threeNumbers(9, 4, 8));
+
+//Task 4
+function area(a, b) {
+  if (b === undefined) return a ** 2;
+  return a * b;
+}
+
+console.log(area(10));
+
+console.log(area(6, 3));
+
+//Task 5
+
+function dividersList(num) {
+  let dividers = [];
+
+  for (let i = Math.trunc(num / 2); i > 0; i--) {
+    if (num % i == 0) {
+      dividers.push(i);
+    }
+  }
+  return dividers;
+}
+
+function perfectNumber(num) {
+  return (
+    num ==
+    dividersList(num).reduce(
+      (accumulator, currentValue) => accumulator + currentValue
+    )
+  );
+}
+
+console.log(perfectNumber(6));
+console.log(perfectNumber(28));
+console.log(perfectNumber(496));
+console.log(perfectNumber(8128));
+console.log(perfectNumber(8129));
+
+//Task 6  v.1
+
+function listGen(start, finish) {
+  let arr = [];
+  for (let i = start; i <= finish; i++) arr.push(i);
+  return arr;
+}
+function allPerfect(start, finish) {
+  let perfectList = [];
+  for (let i = start; i <= finish; i++) {
+    if (perfectNumber(i)) {
+      perfectList.push(i);
+    }
+  }
+  return perfectList;
+}
+console.time("Execution Time");
+console.log(allPerfect(2, 10000));
+console.timeEnd("Execution Time");
+
+//Task 6  v.2 filter method about 10% faster
+
+filterPerfect = (start, finish) =>
+  listGen(start, finish).filter((j) => perfectNumber(j));
+
+console.time("Execution Time");
+console.log(filterPerfect(2, 10000));
+console.timeEnd("Execution Time");
+
```

Comment:
It's good that you use different ways to solve this task but at our lessons I wanted and want people one more time use loop for that because I want people to understand it better
Of course on real life we need to use methods which are already built-in JS